### ISSUE TITLE
OpenAPI and Autorest emitters need updating for shared routes

### ISSUE THREAD

**Issue by tjprescott** [microsoft/cadl#1198]
Issue #1133 enables having multiple operations decorated with `@route` using the `{shared: true}` option on the decorator. The autorest and openAPI emitters should be updated to make use of this feature, as appropriate.

**Comment by markcowl** [comment]
est: 8

**Comment by tjprescott** [comment]
The original design issue cited this example:
```
@route("/uploadImage", { shared: true })
op uploadImageBytes(@body body: bytes, @header contentType: "image/png"): void;

@route("/uploadImage", { shared: true })
op uploadImageJson(@body body: {imageBase64: bytes}, @header contentType: "application/json"): void;
```

**Comment by johanste** [comment]
For swagger/autorest, in order to produce good libraries using autorest, we needed to use x-ms-paths whenever the input parameter combination deterministically determined the return type. Similarly, we needed to use it whenever parameters were mutually exclusive.

I would say that the easiest thing to do is to use x-ms-paths whenever it is easiest for OA2 unless we are already doing the work to disambiguate for OA3 in a way that is OA2 compatible. 

> E.g. any solution that uses anyOf/oneOf or per-content-type responses to solve the problem for OA3 are not OA2 compatible, and thus we can't use them and would fall back to x-ms-paths for OA2.

Doing extra work to produce less precise specs for OA2 to avoid confusing tools that we are not going to use feel like not well-spent time.

**Comment by tjprescott** [comment]
Related issue: https://github.com/Azure/cadl-azure/issues/2438

**Comment by tjprescott** [comment]
Okay, I have these two test cases. The question is whether they should be allowed or disallowed:

**Case 1 - Route contains query info**
```
      @service({title: "My Service"})
      namespace Foo {
        model Resource {
          id: string;
        }
      
        @route("/sharedroutes/resources?filter=resourceGroup")
        op listByResourceGroup(...Resource): Resource[];
      
        @route("/sharedroutes/resources?filter=subscription")
        op listBySubscription(...Resource): Resource[];
      }
```

This is the closest representation to how x-ms-paths works. It should be logical equivalent to the following:

**Case 2 - Shared route with query param constants**
```
      @service({title: "My Service"})
      namespace Foo {
        model Resource {
          id: string;
        }
      
        @route("/sharedroutes/resources", { shared: true})
        op listByResourceGroup(... Resource, @query filter: "resourceGroup"): Resource[];
    
        @route("/sharedroutes/resources", { shared: true})
        op listBySubscription(... Resource, @query filter: "subscription"): Resource[];
```

The problem in case 1 is that trying to simulate the param from case 2 fails because you would have to mock up nodes and all kinds of stuff. It's not easy to just "fake it".

Right now, neither of these scenarios work fully, though there is some logic in place that kinda supports case 1. Should we allow both of these syntaxes, one or neither?

**Comment by johanste** [comment]
I think we should allow full [URI-templates](https://www.rfc-editor.org/rfc/rfc6570) in the `@route` decorator. But that might be a design discussion to be had as there are... implications... there (e.g. how we compute route uniqueness etc.). This would allow for both cases to be supported.

**Comment by tjprescott** [comment]
Right, case 1 would essentially require changes to the `@cadl-lang/rest` package, which I agree should probably require a design discussion. I will say that the current autorest implementation assumed this was possible because it looks for "?" in the path to use "x-ms-paths". 

Case 2 doesn't really require any changes to the rest package, so I can already proceed with making that work. So the question becomes, if I can make Case 2 work, do we also _want_ to support Case 1 or should we proactively disallow case 1 (by making the route decorator validate that it doesn't have query params, for example).

**Comment by markcowl** [comment]
I think full URI-template support would be principled and cool, and a nice feature for Cadl V2.  Representing this in any OpenAPI representation will efinitely require some discussion. I think we should allow both of the above representations, and be explicit about the limits (i.e. we aren't supporting shared path scenarios beyond single query parameters).  This will also allow us to expand into Uri templates in the future, without breaking existing template values.

**Comment by johanste** [comment]
If we just skip openapi 3.1 and move straight to moonwalk, URI templates should be supported :).

**Comment by tjprescott** [comment]
>  I think we should allow both of the above representations, and be explicit about the limits (i.e. we aren't supporting shared path scenarios beyond single query parameter values).

This (case 1) would require a change to the logic in the rest library, and thus might warrant a design discussion. Right now, I am essentially reading the query string and trying to forcibly add the parameter to the method signature, but this blows up because it is not suitably faked (i.e. it doesn't have a node since the param definition didn't actually come from parsed cadl, etc.) If the team is okay with my approach I can keep going, but I don't want to go to far down that rabbit hole only to have it rejected for being too hacky.

**Comment by markcowl** [comment]
@tjprescott Honestly, we need to unblock people by allowing some syntax for this.  If we can implement option 2, let's do that and file a design issue for enabling the additional syntax.

**Comment by markcowl** [comment]
Priority clarification: It would be very nice to have cadl-autorest updated to support this, and nice to have openapi3, however, the required parts of shared routes support are already complete.

**Comment by tjprescott** [comment]
Updated [Playground](https://cadlplayground.z22.web.core.windows.net/?c=aW1wb3J0ICJAdHlwZXNwZWMvaHR0cCI7Cgp1c2luZyBUeXBlU3BlYy5IdHRwOwoKQHNlcnZpY2Uoe3RpdGxlOiAiT3ZlcmxvYWRzJ24gcm91dGVzIn0pCm5hbWVzcGFjZSBFeGFtcGxlczsKCi8vIE9wZW5BUEkzIC0gYWRkIHRoZSBQZW1GaWxlIGNvbnRlbnTICMR6xDJBdXRvcmVzdMcyeC1tcy1wYXRocyAtIHByZXRlbsVHeSBhcmUgZGlmZmVyZW50aWF0ZWQgYnkgcXVlcnkgcGFyYW3rAJVDxmROZWdvxCtpb24gewogIG1vZGVs6QCRxBIgIEBoZWFkZXLsAJc6ICJhcHBsaWPFOi94LXBlbS1maWxlIjvFM2JvZHk6IGJ5dOQA9iAgfcQFx1trY3M3313OXXDEMi1taW3aXQogIEDlAXooIi9nZXRDZXJ0aWbEQGUiLCB7c2hhcmVkOiB0cnVlfSnELGdldAogIG9wIGdldOcA7CjoAIlhY2NlcHQ%2F%2BgDiKTroASDfed956QEK33vrAQDEe8g2Owp9CusB0OkChSB75ACW5gFtUmVzb3VyY%2BgBbGlkOiBzdHJpbmflATjtAL5vyEAvcsc1cyLkALNvcCBsaXN0yEpzKC4uLslYLCBA5gJeZ3JvdXA%2FyF7JF3N1YnNjcmlw5ADIyR4pOslAW13mAUPIfSjNa8t%2FQnnIFkfEavYAhegAucUj327fblPrALnWbewA29ds7QGgU%2BUCDFLlBB%2FlAY8vKuUBlEJyb2tlbiAtIG9ubHkgb25lIG9m5QQV6wO3ZXRlcnMgc2hvd3MgdXAgaW4gb3BlbmFwaSBlbWl0xB0uxkxvdGjEJ3VsZMUuxC1tYXJrZWQgYXMgb%2BUArGFsxizrBFXqBFEKICAqL%2F8CPvUCPuYC6sUO7AJBLCB7IPEDBf8B0%2F8B0%2FYB0%2F8AiP4AiP8B7f8B7fQB7egDjUToBY3yAfhDaGFuY2Vz5QW0dGhhdCB0aGlzIGlzIG1vcuUCBGEgY2xp5AYOc2lkZS1jYXIgbWV0aG9kIHNjZW5hcmlv5AIId2hpY2jFUGNhc2Ug6QPFIHfkAgVuJ3QgYmUgaW52b2x2ZWQsIGJ1dCByYXRoZXIgbXVsdGlwbGUgPG9wIHNvxFzkBu5pc%2BUBFS4uLj7kAVkgIEN1cuQAwWx5IGLoAqppbmNsdWRlc%2BUCp%2BUBbC1sZXNzICJyb290IucApy4gQnV0IHNlZSBhYm92ZeQCdSov%2FwSa7QJc%2FwR5%2FwR5%2FwR58AI46AE56Qc%2B7wJD5QKAcy97zRB97wUr%2FwRZ%2FwL0LP8C7f8AtesAtewA%2FXMve8wP%2FwCz%2FwUM8QMf%2FwMY5gMY)

**Comment by mikekistler** [comment]
Comment on the updated playground: this isn't capturing the multiple response types in OpenAPI 3 correctly.  Looking at just the [ContentNegotiation example](https://cadlplayground.z22.web.core.windows.net/?c=aW1wb3J0ICJAdHlwZXNwZWMvaHR0cCI7Cgp1c2luZyBUeXBlU3BlYy5IdHRwOwoKQHNlcnZpY2Uoe3RpdGxlOiAiT3ZlcmxvYWRzJ24gcm91dGVzIn0pCm5hbWVzcGFjZSBFeGFtcGxlczsKCi8vIE9wZW5BUEkzIC0gYWRkIHRoZSBQZW1GaWxlIGNvbnRlbnTICMR6y0dDxhZOZWdvdGlhdGlvbiB7CiAgbW9kZWzJQ8QSICBAaGVhZGVyzEk6ICJhcHBsaWPFOi94LXBlbS1maWxlIjvFM2JvZHk6IGJ5dOQAqCAgfcQFx1trY3M3313OXXDEMi1taW3aXQogIEDlASwoIi9nZXRDZXJ0aWbEQGUiLCB7c2hhcmVkOiB0cnVlfSnELGdldAogIG9wIGdldOcA7CjoAIlhY2NlcHQ%2F%2BgDiKTroASDfed956QEK33vrAQDEe8g2Owp9Cg%3D%3D), the `content` of the 200 response should have entries for both `application/pkcs7-mime` and `application/x-pem-file`, e.g.:

```
      responses:
        '200':
          description: The request has succeeded.
          content:
            application/pkcs7-mime:
              schema:
                $ref: '#/components/schemas/ContentNegotiation.Pkcs7File'
            application/x-pem-file:
              schema:
                $ref: '#/components/schemas/ContentNegotiation.PemFile'
```
